# Phylogeny Workshop

## Content
1. [Minhash and Neighbour-joining](#minhash-and-neighbour-joining)
2. [IQTREE2 - Maximum Likelihood](#iqtree2---maximum-likelihood)
3. [Bayesian inference (MrBayes)](#bayesian-inference-mrbayes)

## Introduction
In this session, we are going to construct phylogenetic trees in 3 different ways, so you can see the differences in speed and accuracy

### Minhash and Neighbour-joining
Minhash is used to quickly estimate Jaccard similarity between genomic datasets based on k-mers (substrings of fixed length). This distance relationship can be used to construct relational trees

- run this code in the julia REPL

```bash 
julia
using Mashtree
# kmer size = 16, minhash sketch size = 100000
mashtree("/mnt/s-ws/everyone/annotation/", 16, 100000)
```
- How long did it take?
- Quit Julia with Ctrl-D

- you can rename the sequences to add the NCBI taxid
- replace xxx with your Newick string

```bash
sourcedir=/mnt/s-ws/everyone/annotation/

tree="xxx"

while IFS=$'\t' read -r latin name taxid accession carnivorous australian
do
    tree=$(sed "s/$accession/$taxid/g" <<< $tree)
done < <(tail -n +2 $sourcedir/genome_metadata.tsv)
echo $tree
```

- paste the Newick tree into https://itol.embl.de/

- Root the tree with Liriodendron tulipifera
- What species is Aldrovanda most closely related to in this tree?
- Try different values for the kmer size (10√¢‚Ç¨‚Äú32) and the sketch size (1000√¢‚Ç¨‚Äú100000)
- Do you get different trees?


### IQTREE2 - Maximum Likelihood
IQTREE2 is a versatile software tool used for constructing phylogenetic trees based on maximum likelihood (ML) methods. It employs codon models for nucleotide sequence evolution, making it suitable for analyzing protein-coding genes

- First we will make a codon alignment using the multiple sequence alignment of
- proteins you made in the 'Genome Annotation' module to guide the alignment of the
- corresponding nucleotide sequences

```bash
cat *.rpoC2.nt.fa > allrpoC2.nt.fa
#the below calls to run the julia script
julia /mnt/s-ws/everyone/tools/nt2codon.jl rpoC2.protein.msa allrpoC2.nt.fa rpoC2.codon.msa
less rpoC2.codon.msa
```

- Now we can use iqtree2 to construct a phylogenetic tree by maximum likelihood using this data.
- `-T 16` tells iqtree2 to use 16 parallel 'threads' so that it can use multiple processors in the server simultaneously.
- `-st` CODON tells iqtree2 to use codon models for sequence evolution.
- `-o` is telling iqtree2 what the outgroup taxon is so it can root the trees correctly.
- `--alrt 1000` and `-B 1000` tell iqtree2 to estimate the confidence in the branches in the tree using 1000 replicates using an approximate likelihood ratio test (--alrt) and a conventional bootstrap test (-B). Much more info on iqtree2 here: http://www.iqtree.org/doc/

- We're going to run iqtree in a batch queue to make sure the server can handle the load

```bash
batch

iqtree2 -T 16 -s rpoC2.codon.msa -st CODON -o NC_008326.rpoC2 --alrt 1000 -B 1000
```

- Ctrl-D to get out of batch
- To see where you are in the queue:

```bash
atq
```

- you can use top to see what the server is doing

```bash
top
```

- type `q` to get out of top

- when `atq` no longer shows you in the queue, iqtree has finished

```bash
less rpoC2.codon.msa.log
```

- How long did iqtree take (CPU time is what matters)?

```bash
less rpoC2.codon.msa.iqtree
```

- paste the Newick tree into https://itol.embl.de/
- Is the tree the same or different to the one generated by Mashtree?


### Bayesian inference (MrBayes)

- Use the same data as for iqtree2, but it needs to be converted to nexus format

```bash
julia /mnt/s-ws/everyone/tools/fasta2nex.jl rpoC2.codon.msa rpoC2.codon.nex

less rpoC2.codon.nex
```

- Run MrBayes interactively

```bash
mb

exe rpoC2.codon.nex
```

- the `lset` command sets the constraints on the sequence evolution model. Here we are using very permissive settings, so MrBayes will test many different possible parameters

```bash
lset nucmodel=codon omegavar=M3 nst=mixed rates=invgamma
```

- The following commands set the frequency with which MrBayes provides feedback on progress; as this run is just to get an idea of how the Bayesian approach works, we are asking for more frequent updates than would be typical

```bash
mcmcp ngen=10000

mcmcp diagnfreq=1000

mcmcp samplefreq=100

mcmcp printfreq=100
```

- Now we start the run proper

```bash
mcmc
```
- Once the run has finished, we can inspect the output

```bash
sump

sumt

cat rpoC2.codon.nex.con.tre 
```

- paste the Nexus tree into https://itol.embl.de/

- Which of the three methods is the fastest? Which is the most accurate?

- Let's take a look at the correct tree
- Use less to see the tree in Newick format

```bash
less $sourcedir/carnivores.nw
```
- paste the tree into https://itol.embl.de/ to visualise it in a graphical form
- Root the tree with *Liriodendron*
- What species is *Aldrovanda* most closely related to?

üå±üîçüìä**Congratulations on completing the Phylogenetics lab!** üåøüî¨‚ú®**or at least you got to the end of this session** üéì

